sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java

```diff
@@ -1542,12 +1542,26 @@ public void startBundle(StartBundleContext c) {
       }
     }
 
+    private static com.google.datastore.v1.Key getKey(Mutation m) {
+      if (m.hasUpsert()) {
+        return m.getUpsert().getKey();
+      } else if (m.hasInsert()) {
+        return m.getInsert().getKey();
+      } else if (m.hasDelete()) {
+        return m.getDelete();
+      } else if (m.hasUpdate()) {
+        return m.getUpdate().getKey();
+      } else {
+        throw new IllegalArgumentException(String.format("Mutation %s is not valid.", m));
```

Comment:
Are these 4 cases exhaustive? I also see `OPERATION_NOT_SET` in Mutation's source code. Maybe just add a warning and return an empty Key? Otherwise LGTM.

Reply:
ok I've done that. Thanks Yi!